File: agent/tools/asset_lookup.go
package tools

import (
	"fmt"
	"strings"

	"github.com/vultisig/vultisig-win/agent/shared"
)

type AssetLookupTool struct{}

func NewAssetLookupTool() *AssetLookupTool {
	return &AssetLookupTool{}
}

func (t *AssetLookupTool) Name() string {
	return "asset_lookup"
}

func (t *AssetLookupTool) Description() string {
	return "Resolve an asset alias to its chain and token information. Supports aliases like 'eth', 'usdc', 'usdc:arbitrum', etc."
}

func (t *AssetLookupTool) InputSchema() map[string]any {
	return map[string]any{
		"asset": map[string]any{
			"type":        "string",
			"description": "Asset alias (e.g., 'eth', 'usdc', 'usdc:arbitrum', 'btc')",
		},
	}
}

func (t *AssetLookupTool) RequiresPassword() bool {
	return false
}

func (t *AssetLookupTool) RequiresConfirmation() bool {
	return false
}

func (t *AssetLookupTool) Execute(input map[string]any, ctx *ExecutionContext) (any, error) {
	assetRaw, ok := input["asset"]
	if !ok {
		return nil, fmt.Errorf("asset is required")
	}

	assetStr := strings.TrimSpace(assetRaw.(string))
	asset := shared.ResolveAsset(assetStr)

	if asset == nil {
		availableAssets := make([]string, 0)
		for k := range shared.AssetAliases {
			availableAssets = append(availableAssets, k)
		}
		return map[string]any{
			"found":            false,
			"input":            assetStr,
			"message":          fmt.Sprintf("Asset '%s' not found in known aliases", assetStr),
			"available_assets": availableAssets,
			"ui": map[string]any{
				"title":   "Asset Lookup",
				"summary": fmt.Sprintf("No asset found for %s", assetStr),
			},
		}, nil
	}

	return map[string]any{
		"found":         true,
		"input":         assetStr,
		"chain":         asset.Chain,
		"ticker":        asset.Ticker,
		"token_address": asset.TokenAddress,
		"is_native":     asset.IsNative,
		"decimals":      asset.Decimals,
		"ui": map[string]any{
			"title":   "Asset Lookup",
			"summary": fmt.Sprintf("%s resolves to %s on %s", assetStr, asset.Ticker, asset.Chain),
			"actions": []map[string]any{
				{
					"type":  "copy",
					"label": "Copy Ticker",
					"value": asset.Ticker,
				},
			},
		},
	}, nil
}

File: core/ui/agent/tools/handlers/assetLookup.ts
import { assetAliases, resolveAsset } from '../shared/assetResolution'
import type { ToolHandler } from '../types'

export const handleAssetLookup: ToolHandler = async input => {
  const assetStr = String(input.asset ?? '').trim()
  if (!assetStr) throw new Error('asset is required')

  const asset = resolveAsset(assetStr)

  if (!asset) {
    const availableAssets = Object.keys(assetAliases)
    return {
      data: {
        found: false,
        input: assetStr,
        message: `Asset '${assetStr}' not found in known aliases`,
        available_assets: availableAssets,
        ui: {
          title: 'Asset Lookup',
          summary: `No asset found for ${assetStr}`,
        },
      },
    }
  }

  return {
    data: {
      found: true,
      input: assetStr,
      chain: asset.chain,
      ticker: asset.ticker,
      token_address: asset.tokenAddress,
      is_native: asset.isNative,
      decimals: asset.decimals,
      logo: asset.logo,
      price_provider_id: asset.priceProviderId,
      ui: {
        title: 'Asset Lookup',
        summary: `${assetStr} resolves to ${asset.ticker} on ${asset.chain}`,
        actions: [{ type: 'copy', label: 'Copy Ticker', value: asset.ticker }],
      },
    },
  }
}
